src/core/utils/layerutils.cpp

```diff
@@ -312,21 +312,46 @@ QgsFeature LayerUtils::duplicateFeature( QgsVectorLayer *layer, const QgsFeature
   } );
   auto sweaper = qScopeGuard( [layer, connection] { layer->disconnect( connection ); } );
 
-  duplicatedFeature = QgsVectorLayerUtils::createFeature( layer, feature.geometry(), feature.attributes().toMap() );
-  if ( layer->addFeature( duplicatedFeature ) )
+  QgsVectorLayerUtils::QgsDuplicateFeatureContext duplicateFeatureContext;
+  duplicatedFeature = QgsVectorLayerUtils::duplicateFeature( layer, feature, QgsProject::instance(), duplicateFeatureContext );
+  const auto duplicateFeatureContextLayers = duplicateFeatureContext.layers();
+
+  // commit changes
+  if ( !layer->commitChanges() )
   {
-    // commit changes
-    if ( !layer->commitChanges() )
+    const QString msgs = layer->commitErrors().join( QStringLiteral( "\n" ) );
+    QgsMessageLog::logMessage( tr( "Cannot add new feature in layer \"%1\". Reason:\n%2" ).arg( layer->name(), msgs ), "QField", Qgis::Warning );
+
+    for ( QgsVectorLayer *chl : duplicateFeatureContextLayers )
     {
-      const QString msgs = layer->commitErrors().join( QStringLiteral( "\n" ) );
-      QgsMessageLog::logMessage( tr( "Cannot add new feature in layer \"%1\". Reason:\n%2" ).arg( layer->name(), msgs ), "QField", Qgis::Warning );
-      return QgsFeature();
+      chl->rollBack();
     }
+
+    return QgsFeature();
   }
-  else
+
+  // we have to re-apply referenced feature attribute value to take into account value changes form the
+  // data provider (e.g. autogenerated fid)
+  const QList<QgsRelation> relations = QgsProject::instance()->relationManager()->referencedRelations( layer );
+  for ( QgsVectorLayer *chl : duplicateFeatureContextLayers )
   {
-    QgsMessageLog::logMessage( tr( "Cannot add new feature in layer \"%1\"." ).arg( layer->name() ), "QField", Qgis::Warning );
-    return QgsFeature();
+    const QgsFeatureIds fids = duplicateFeatureContext.duplicatedFeatures( chl );
+    for ( auto relation : relations )
+    {
+      if ( relation.strength() == Qgis::RelationshipStrength::Composition && relation.referencingLayer() == chl )
```

Comment:
I don't understand this completely, why does the relationship strength matter here? Can't we just update all fids from the context without checking any relationship strength here?

Reply:
True